src/i2c.rs

```diff
@@ -0,0 +1,369 @@
+//! Inter-Integrated Circuit (I2C) bus
+
+use afio::MAPR;
+use gpio::{Alternate, OpenDrain};
+use gpio::gpiob::{PB10, PB11, PB6, PB7, PB8, PB9};
+use hal::blocking::i2c::{Read, Write, WriteRead};
+use rcc::{APB1, Clocks};
+use stm32f103xx::{I2C1, I2C2};
+
+/// I2C error
+#[derive(Debug)]
+pub enum Error {
+    /// Timed out busy waiting
+    Timeout,
+    /// Bus error
+    Bus,
+    /// Arbitration loss
+    Arbitration,
+    /// No ack received
+    Acknowledge,
+    /// Overrun/underrun
+    Overrun,
+    // Pec, // SMBUS mode only
+    // Timeout, // SMBUS mode only
+    // Alert, // SMBUS mode only
+    #[doc(hidden)] _Extensible,
+}
+
+#[derive(Debug)]
+pub enum DutyCycle {
+    Ratio1to1,
+    Ratio16to9,
+}
+
+#[derive(Debug)]
+pub enum Mode {
```

Comment:
This looks like it lets you create invalid states. For example, `Mode::Standard { frequency: 400_000 }` would no longer be the standard mode if the frequency is higher than 100 KHz (at least that's my recollection of the spec).

I think something similar happened with the duty cycle: once you are above some frequency you are supposed to use the 16:9 duty cycle, iirc.

In the f30x-hal I simply let the caller pass a frequency and then I pick the duty cycle based on what mode they implicitly requested.

Reply:
Well, that might work in your case but the I2C specification states limits for the normal, fast, etc. modes. I'm of the opinion that we should stick to the standard here and not allow 1 Mbps standard mode, etc.